File: packages/python/tests/integration/interpret_subprocess.py
"""Subprocess plumbing for the `dirsql interpret` integration tests.

Kept in its own module (rather than `conftest.py` or inline in the
test file) because these helpers are not pytest fixtures -- they are
plain functions that drive a child process via NDJSON. Importing them
keeps each test focused on the assertion, not the framing.

Naming intentionally does not match `test_*.py` / `*_test.py` so
pytest skips collection.
"""

from __future__ import annotations

import json
import shutil
import subprocess
import threading
from pathlib import Path
from typing import Any


def cli_argv() -> list[str]:
    """Resolve the `dirsql` console-script invocation for this test env.

    Failing loudly here surfaces an environment misconfiguration rather
    than masking it as a test assertion failure further down.
    """
    dirsql = shutil.which("dirsql")
    assert dirsql is not None, (
        "`dirsql` console script not on PATH -- run `uv run maturin develop`"
    )
    return [dirsql]


def spawn(config_path: Path) -> subprocess.Popen:
    """Start `dirsql interpret <config>` with piped stdin/stdout/stderr."""
    return subprocess.Popen(
        [*cli_argv(), "interpret", str(config_path)],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        bufsize=1,
    )


def readline(proc: subprocess.Popen, timeout: float = 5.0) -> str:
    """Read one stdout line with a timeout.

    On timeout, kill the process and surface stderr so the failure
    message is actionable rather than an opaque hang.
    """
    result: list[str | None] = [None]

    def reader() -> None:
        assert proc.stdout is not None
        result[0] = proc.stdout.readline()

    t = threading.Thread(target=reader, daemon=True)
    t.start()
    t.join(timeout)
    if t.is_alive():
        proc.kill()
        stderr = proc.stderr.read() if proc.stderr else ""
        raise AssertionError(
            f"timed out waiting for stdout line; stderr was:\n{stderr}"
        )
    line = result[0] or ""
    if not line:
        stderr = proc.stderr.read() if proc.stderr else ""
        raise AssertionError(
            f"helper exited (code={proc.returncode}) before writing a line; "
            f"stderr was:\n{stderr}"
        )
    return line


def send(proc: subprocess.Popen, msg: dict[str, Any]) -> None:
    """Write one NDJSON line to the helper's stdin."""
    assert proc.stdin is not None
    proc.stdin.write(json.dumps(msg) + "\n")
    proc.stdin.flush()


def shutdown(proc: subprocess.Popen) -> None:
    """Close stdin and wait for the helper to exit cleanly."""
    if proc.stdin is not None:
        try:
            proc.stdin.close()
        except BrokenPipeError:
            pass
    try:
        proc.wait(timeout=5)
    except subprocess.TimeoutExpired:
        proc.kill()
        proc.wait(timeout=2)

File: packages/ts/tests/e2e/interpret-subprocess.ts
// Subprocess plumbing for the `dirsql interpret` integration tests.
//
// Kept in its own module (not a `.test.ts` file) so vitest doesn't try
// to collect it. Imports are runtime-only; nothing here is part of the
// shipped TS SDK.

import { type ChildProcessWithoutNullStreams, spawn } from "node:child_process";
import { access } from "node:fs/promises";
import {
  type Interface as ReadlineInterface,
  createInterface,
} from "node:readline";

export interface InterpretHandle {
  proc: ChildProcessWithoutNullStreams;
  lines: AsyncIterator<string>;
  rl: ReadlineInterface;
  stderrChunks: string[];
}

/** Start `<node> <cliEntry> interpret <configPath>` with piped streams. */
export async function spawnInterpret(
  cliEntry: string,
  configPath: string,
  opts: { cwd?: string } = {},
): Promise<InterpretHandle> {
  try {
    await access(cliEntry);
  } catch {
    throw new Error(
      `CLI entry not built: ${cliEntry} -- run \`pnpm build\` first`,
    );
  }
  const proc = spawn(process.execPath, [cliEntry, "interpret", configPath], {
    stdio: ["pipe", "pipe", "pipe"],
    cwd: opts.cwd,
  });
  const rl = createInterface({ input: proc.stdout });
  const stderrChunks: string[] = [];
  proc.stderr.setEncoding("utf8");
  proc.stderr.on("data", (chunk: string) => stderrChunks.push(chunk));
  return { proc, lines: rl[Symbol.asyncIterator](), rl, stderrChunks };
}

/** Read one stdout line; reject with a stderr-bearing error on timeout/EOF. */
export async function readLine(
  handle: InterpretHandle,
  timeoutMs = 5_000,
): Promise<string> {
  const next = handle.lines.next();
  const timer = new Promise<{ done: true; value: undefined }>(
    (_resolve, reject) => {
      setTimeout(
        () =>
          reject(
            new Error(
              `timed out reading line; stderr: ${handle.stderrChunks.join("")}`,
            ),
          ),
        timeoutMs,
      );
    },
  );
  const result = (await Promise.race([next, timer])) as IteratorResult<string>;
  if (result.done) {
    throw new Error(
      `helper exited before writing a line; stderr: ${handle.stderrChunks.join("")}`,
    );
  }
  return result.value;
}

/** Write one NDJSON line to the helper's stdin. */
export function send(handle: InterpretHandle, msg: unknown): void {
  handle.proc.stdin.write(`${JSON.stringify(msg)}\n`);
}

/** Close stdin and wait for the helper to exit; SIGKILL after 5s. */
export async function shutdown(handle: InterpretHandle): Promise<void> {
  handle.rl.close();
  handle.proc.stdin.end();
  await new Promise<void>((resolve) => {
    if (handle.proc.exitCode !== null || handle.proc.signalCode !== null) {
      resolve();
    } else {
      handle.proc.once("close", () => resolve());
      setTimeout(() => {
        if (handle.proc.exitCode === null && handle.proc.signalCode === null) {
          handle.proc.kill("SIGKILL");
        }
      }, 5_000);
    }
  });
}
